src/python/test/test_extract_dataset.py

```diff
@@ -0,0 +1,134 @@
+               "patient.date_of_birth", ",", "patient.verified_dtm", ",", "test.test_id", ",", "trial_visit.visit_id", ",", "baseline.baseline_id", 
+               ",", "hospital.hospital_id", ",", "doctor.doctor_id","--sql", "-o", "-"]
+        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, universal_newlines=True)
+        stdout = process.communicate()[0]
+        self.assertTrue(truth_output==stdout.strip())
+
+    def test_e2e_dataset_fields(self):
+        dataset_record = "project-G9j1pX00vGPzF2XQ7843k2Jq:record-GBGpKZ80G8j8k4G84P7GP0B5"
+        out_directory = tempfile.mkdtemp()
+        cmd = ["dx", "extract_dataset", dataset_record, "--fields", "patient.patient_id" , ",", "patient.name", ",", "patient.weight", ",",
+               "patient.date_of_birth", ",", "patient.verified_dtm", ",", "test.test_id", ",", "trial_visit.visit_id", ",", "baseline.baseline_id", 
+               ",", "hospital.hospital_id", ",", "doctor.doctor_id", "-o", out_directory]
+        subprocess.check_call(cmd)
+        truth_file = "project-G9j1pX00vGPzF2XQ7843k2Jq:file-GBGpx800vGPy4qf84JJQBX3x"
+        self.end_to_end_fields(out_directory=out_directory, rec_name = "extract_dataset_test.csv", truth_file=truth_file)
+
+    def test_e2e_cohortbrowser_fields(self):
+        cohort_record = "project-G9j1pX00vGPzF2XQ7843k2Jq:record-GBGq1pQ0vGPf0qYg30Jg7kkG"
+        out_directory = tempfile.mkdtemp()
+        cmd = ["dx", "extract_dataset", cohort_record, "--fields", "patient.patient_id" , ",", "patient.name", ",", "patient.weight", ",",
+               "patient.date_of_birth", ",", "patient.verified_dtm", ",", "test.test_id", ",", "trial_visit.visit_id", ",", "baseline.baseline_id", 
+               ",", "hospital.hospital_id", ",", "doctor.doctor_id", "-o", out_directory]
+        subprocess.check_call(cmd)
+        truth_file = "project-G9j1pX00vGPzF2XQ7843k2Jq:file-GBGq7FQ0vGPpK3GbF0Xbjz17"
+        self.end_to_end_fields(out_directory=out_directory, rec_name = "Combined_Cohort_Test.csv", truth_file=truth_file)
+    
+    def test_file_already_exists(self):
+        cohort_record = "project-G9j1pX00vGPzF2XQ7843k2Jq:record-GBGq1pQ0vGPf0qYg30Jg7kkG"
+        out_directory = tempfile.mkdtemp()
+        open(os.path.join(out_directory, "Combined_Cohort_Test.csv"), 'w').close()
+        cmd = ["dx", "extract_dataset", cohort_record, "--fields", "patient.patient_id" , ",", "patient.name", "-o", out_directory]
+        output = subprocess.run(cmd, capture_output=True, text=True)
+        self.assertTrue("Error: path already exists" in output.stderr)
+        shutil.rmtree(out_directory)
+
+    def end_to_end_ddd(self, out_directory, rec_name):
+        truth_files_directory = tempfile.mkdtemp()
+        os.chdir(truth_files_directory)
+        cmd = ["dx", "download", "project-G9j1pX00vGPzF2XQ7843k2Jq:file-GBGp61j0vGPpz58XK9f7Vkj6", 
```

Comment:
Is the project public (on staging)?

We try to keep dx-toolkit test environment-agnostic (prod/stg-agnostic) so is it possible to create the content of the files on the fly as the tests are run (and use it for the assertion/truth)?

Reply:
I don't think the project is public. May be @kpjensen can tell us if this will be made public. `user-jenkins` has view access in the project for the purpose of jenkins build and test.

The input to `extract_dataset` is a dataset or cohort on staging/platform. If this has to be created on the fly, we will need to check-in `data_dictionary.csv`, `entity_dictionary.csv`, `codings.csv` and the`test_data_to_load.csv` to the repo. Then run `data model loader`(DML) to create the dataset and use it in extract_dataset. Running DML can take around 5 minutes to create a small dataset.

After running DML and extract dataset, for validating the results we need truth files. Those will also have to be checked in the repo.

Reply:
Not planning to make the staging project public. @commandlinegirl given that this functionality is isolated from the rest of dx-toolkit I'm okay with the current approach, although perhaps we should set up the same data in prod and then change these tests to refer to `project_name:filename` instead of object IDs.

Reply:
Updated.